### PR TITLE
Incremental Supervisor improvements

### PR DESCRIPTION
**Pull request by jsiwek**
* Improved stdout/stderr handling: the Supervisor now intercepts streams from all children and even invokes script-level hook to allow for customized logic

* Handling of leftover log files: in the event of a crash/kill, unrotated logs are now detected and a rotate/postprocess attempted at startup-time

* New mechanisms to control log file rotation formatting (e.g. timestamp style) and rotation directory (e.g. helps move toward goal of using external "archiver" process to monitor a directory and archive its contents... implementation for that will come separately later on)

* Added ability to compare Zeek functions for equality: needed in pursuit of implementing the new log file rotation options and simplifying the existing ASCII rotation postprocessing logic in a backwards-compatible manner.

Additional PR that goes with this: https://github.com/zeek/zeek-docs/pull/29

### CODE REVIEW

File: scripts/base/frameworks/logging/main.zeek [hunk 1]
```diff
@@ -119,13 +119,62 @@ export {
...
+
+		## A prefix to use for the the rotated log.  Log writers may later
+		## append a file extension of their choosing to this user-chosen
+		## prefix (e.g. if using the default ASCII writer and you want
+		## rotated files of the format "foo-<date>.log", then this prefix
+		## can be set to "foo-<date>" and the ".log" is added later (there's
+		## also generally means of customizing the file extension, too,
+		## like the ``ZEEK_LOG_SUFFIX`` environment variable or
+		## writer-dependent configuration options.
+		file_prefix: string;
```
Comment (rsmmr) [inline scripts/base/frameworks/logging/main.zeek:169]: I find `prefix` a bit misleading here, how about `file_basename` instead?

File: scripts/base/frameworks/logging/main.zeek [hunk 2]
```diff
@@ -119,13 +119,62 @@ export {
...
+	option default_rotation_dir = "";
+
+	## A log file rotation path specification that's returned by the
+	## user-customizable :zeek:see:`Log::rotation_format_func`.
+	type RotationPath: record {
+		## A directory to rotate the log to.  This directory is created
+		## just-in-time, as the log rotation is about to happen.  If it
+		## cannot be created, an error is emitted and the rotation process
+		## tries to proceed with rotation inside the working directory.  When
+		## setting this field, beware that renaming files across systems will
```
Comment (rsmmr) [inline scripts/base/frameworks/logging/main.zeek:157]:
```suggestion
		## setting this field, beware that renaming files across file systems will
```

File: scripts/base/frameworks/supervisor/api.zeek [hunk 3]
```diff
@@ -117,4 +117,28 @@ export {
 	##          It's an error to call this function from a process other than
 	##          a supervised one.
 	global node: function(): NodeConfig;
+
+	## Hooks intercepted stdout stream for all supervisor's child processes.
```
Comment (rsmmr) [inline scripts/base/frameworks/supervisor/api.zeek:121]:
```suggestion
	## Hooks intercepted stdout stream for all supervisor's child processes.
	## If a hook terminates with `break`, that will suppress output to the associated stream.
```

File: scripts/base/frameworks/supervisor/api.zeek [hunk 4]
```diff
@@ -117,4 +117,28 @@ export {
...
+	##       :zeek:see:`Supervisor::create` indicating to which
+	##       child process the stdout line is associated.
+	##       An empty value is used to indicate the message
+	##       came from the internal supervisor stem process
+	##       (this should typically never happen).
+	##
+	## msg: line-buffered contents from the stdout of a child process.
+	global stdout_hook: hook(node: string, msg: string);
+
+	## Hooks intercepted stderr stream for all supervisor's child processes.
```
Comment (rsmmr) [inline scripts/base/frameworks/supervisor/api.zeek:133]:
```suggestion
	## Hooks intercepted stderr stream for all supervisor's child processes.
	## If a hook terminates with `break`, that will suppress output to the associated stream.
```

File: src/logging/Manager.cc [hunk 5]
```diff
@@ -1493,22 +1498,78 @@ std::string Manager::FormatRotationTime(time_t t)
...
+                                        zeek::FuncPtr postprocessor)
 	{
-	auto rot_str = FormatRotationTime(t);
-	return fmt("%.*s-%s",
-	           static_cast<int>(path.size()), path.data(), rot_str.data());
+	auto ri = zeek::make_intrusive<zeek::RecordVal>(zeek::BifType::Record::Log::RotationFmtInfo);
+	ri->Assign(0, std::move(writer));
+	ri->Assign<zeek::TimeVal>(2, open);
+	ri->Assign<zeek::StringVal>(1, path.size(), path.data());
+	ri->Assign<zeek::TimeVal>(3, close);
```
Comment (rsmmr) [inline src/logging/Manager.cc:1510]:
```suggestion
	ri->Assign<zeek::StringVal>(1, path.size(), path.data());
	ri->Assign<zeek::TimeVal>(2, open);
	ri->Assign<zeek::TimeVal>(3, close);
```

File: src/logging/writers/ascii/Ascii.cc [hunk 6]
```diff
@@ -753,50 +739,61 @@ void Ascii::RotateLeftoverLogs()
...
+		auto ppf = default_ppf;
+
+		if ( ! ll.post_proc_func.empty() )
+			{
+			auto func = zeek::id::find_func(ll.post_proc_func.data());
+
+			if ( func )
+				ppf = std::move(func);
+			else
+				reporter->Warning("Could no postprocess log '%s' with intended "
```
Comment (rsmmr) [inline src/logging/writers/ascii/Ascii.cc:758]:
```suggestion
				reporter->Warning("Could not postprocess log '%s' with intended "
```

File: src/logging/writers/ascii/Ascii.cc [hunk 7]
```diff
@@ -16,6 +28,146 @@ using namespace threading;
...
+	 * trying to parse the shadow file and construct an object.
+	 */
+	std::string error;
+
+	/**
+	 * Return the name of the log without file extension (this is
+	 * what's called "path" in other logging framework parlance).
+	 * E.g. the name of "conn.log" is just "conn".
+	 */
+	std::string Name() const
```
Comment (rsmmr) [inline src/logging/writers/ascii/Ascii.cc:88]: Any reason not to call this `Path` then?

File: src/logging/writers/ascii/Ascii.cc [hunk 8]
```diff
@@ -16,6 +28,146 @@ using namespace threading;
...
+	auto sf_len = ftell(sf_stream);
+	fseek(sf_stream, 0, SEEK_SET);
+
+	auto sf_content = std::make_unique<char[]>(sf_len);
+	auto bytes_read = fread(sf_content.get(), 1, sf_len, sf_stream);
+
+	if ( bytes_read != static_cast<size_t>(sf_len) )
+		{
+		rval.error = "Failed to read contents of " + rval.shadow_filename;
+		return rval;
```
Comment (rsmmr) [inline src/logging/writers/ascii/Ascii.cc:126]:
```suggestion
		fclose(sf_stream);
		return rval;
```

File: src/logging/writers/ascii/Ascii.cc [hunk 9]
```diff
@@ -16,6 +28,146 @@ using namespace threading;
...
+		}
+
+	fclose(sf_stream);
+	std::string_view sf_view(sf_content.get(), sf_len);
+	auto sf_lines = tokenize_string(sf_view, '\n');
+
+	if ( sf_lines.size() < 2 )
+		{
+		snprintf(errbuf, sizeof(errbuf),
```
Comment (rsmmr) [inline src/logging/writers/ascii/Ascii.cc:135]: Just use `fmt`?

File: src/supervisor/Supervisor.cc [hunk 10]
```diff
@@ -626,29 +761,32 @@ std::optional<Supervisor::SupervisedNode> Stem::Revive()
...
-		LogError("Supervised node '%s' (PID %d) revived after premature exit",
-		         node.Name().data(), node.pid);
 		ReportStatus(node);
 		}
 
 	return {};
 	}
 
-std::optional<Supervisor::SupervisedNode> Stem::Spawn(Supervisor::Node* node)
+std::variant<bool, Supervisor::SupervisedNode> Stem::Spawn(Supervisor::Node* node)
```
[the remark this answers is not in the document]
Comment (jsiwek) [reply src/supervisor/Supervisor.cc]: Will add a comment:
```
	/**
	 * This performs fork() to initialize the supervised-node structure.
	 * There's three possible outcomes:
	 *   - return value is SupervisedNode: we are the child process
	 *   - return value is True: we are the parent and fork() succeeded
	 *   - return value is False: we are the parent and fork() failed
	 */
```
There was one place where the distinction between passing/failing `fork()` mattered.

File: scripts/base/frameworks/logging/main.zeek (hunk 1 above)
Comment (jsiwek) [reply scripts/base/frameworks/logging/main.zeek:169]: Sure, changed.

File: src/logging/writers/ascii/Ascii.cc (hunk 9 above)
Comment (jsiwek) [reply src/logging/writers/ascii/Ascii.cc:135]: Sure (guessing this originally lived within the thread where that wasn't safe to do).

File: src/logging/writers/ascii/Ascii.cc (hunk 7 above)
Comment (jsiwek) [reply src/logging/writers/ascii/Ascii.cc:88]: The reason was probably just that the "path" terminology of the logging framework doesn't match my expectations of it being related to real filesystem paths, but I'll change it to `Path()` for consistency.

File: src/supervisor/Supervisor.cc [hunk 11]
```diff
@@ -396,9 +488,50 @@ double Supervisor::GetNextTimeout()
...
+	buffer.clear();
+	pipe = nullptr;
+	}
+
+size_t zeek::detail::LineBufferedPipe::Process()
+	{
+	if ( ! pipe )
+		return 0;
+
+	char buf[256];
```
[the remark this answers is not in the document]
Comment (jsiwek) [reply src/supervisor/Supervisor.cc]: Factored.

**Comment by jsiwek** [comment]
@rsmmr thanks for review/suggestions, addressed+merged

**Pull request merged**